### PR TITLE
Introduce layouts (formerly widgets) to prepare for redesign

### PR DESCRIPTION
**Pull request by mmilata**
Most changes in this PR come from #1312, except there are no UI bits for T1, and the layouts have been tweaked to pass most UI tests.

Changes:
* UI components moved to model-specific directories, for some of them the model-independent code has been factored out
* some dialogs have been converted to the `Layout` abstraction (formerly `Widget`)

To be done:
* convert all  `app.*` code to use `Layout`s, remove direct imports of `trezor.ui.components.tt.*`
* update `ButtonRequest` message to include string field "type" (currently it is just written to log)

There are some spacing differences in the [UI diff](https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/956622686/artifacts/master_diff/index.html), if these are a problem they can probably be worked around.

### CODE REVIEW

File: core/src/apps/tezos/get_public_key.py
Comment (matejcik) [inline core/src/apps/tezos/get_public_key.py:4]: couldn't we use `show_pubkey` here too?

File: core/src/trezor/ui/components/tt/confirm.py
Comment (matejcik) [inline core/src/trezor/ui/components/tt/confirm.py:34]: please use the new-style type annotations: `button_confirm: Optional[Button] = None`

File: core/src/trezor/ui/layouts/tt.py
Comment (matejcik) [inline core/src/trezor/ui/layouts/tt.py:187]: please use the new-style type annotation

**Review by matejcik — CHANGES_REQUESTED** [review]
minor comments

overall: why did you choose to have `show_pubkey` accept str instead of bytes?
not saying it's wrong (i did the same in `confirm_hex` or something), just wondering about reasoning and whether in case of `show_pubkey` the hexlification could be done in the layout; we _know_ that pubkey is bytes

otherwise basically LGTM, let's wait until the release branch, and then rebase&merge

File: core/src/apps/tezos/get_public_key.py
Comment (mmilata) [reply core/src/apps/tezos/get_public_key.py]: Not sure if I overlooked it or didn't use it because of slight difference ("..." at the end if pubkey doesn't fit screen). We definitely should unify those (and make them scrollable if the pubkey doesn't fit), fixed in 48190477e1de3bf7ceca76880c7dcc587819fe42.

File: core/src/trezor/ui/components/tt/confirm.py
Comment (mmilata) [reply core/src/trezor/ui/components/tt/confirm.py]: Fixed in 81b9c568ba0a0f3e9e070a6d56990094d2937857. `# type: ignore` can't be expressed in the new style, right?

File: core/src/trezor/ui/layouts/tt.py
Comment (mmilata) [reply core/src/trezor/ui/layouts/tt.py]: Fixed in 379440b0a6baa70f6bebf4542e3fdee21c6f092c.

**Comment by mmilata** [comment]
Addressed the comments and rebased to master. Conflicts:
* `apps/bitcoin/get_address.py` (xpub magic, "others" -> "cosigner")
* `apps/bitcoin/sign_tx/layout.py` (amount units)
* `apps/cardano/layout.py` (map account paths to account numbers) - rolled changes back to what's in master

File: core/src/trezor/ui/components/tt/confirm.py
Comment (matejcik) [reply core/src/trezor/ui/components/tt/confirm.py]: correct; it's not a type, it's an instruction to the typechecker

**Comment by mmilata** [comment]
Rebased on master to resolve conflicts in fixtures. Ready to merge? [UI diff](https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/1015469580/artifacts/master_diff/index.html)

**Review by matejcik — APPROVED** [review]
approval from me

but pinging @tsusanka, this is big enough that i'd like another pair of eyes

**Review by tsusanka — APPROVED** [review]
LGTM! I have skimmed through the code and discussed few things on-call with @mmilata. Let's merge it! :tada:

**Review by tsusanka — APPROVED** [review]
LGTM! I have skimmed through the code and discussed few things with @mmilata on-call. Let's merge this!

**Review by tsusanka — APPROVED** [review]
LGTM! I have skimmed through the code and discussed few things with @mmilata on-call. Let's merge this!

**Comment by tsusanka** [comment]
Due to Github's 500 I have approved 3x times :roll_eyes:.

**Comment by mmilata** [comment]
Thanks a lot for reviewing this thing guys! @matejcik can you please look at this little fix: f9f91a2b71a82d0d369e2339c5f64c86c9660e18?  Without it the Cancel button on get_public_key confirm doesn't actually cancel the operation.

**Comment by matejcik** [comment]
> Thanks a lot for reviewing this thing guys! @matejcik can you please look at this little fix: [f9f91a2](https://github.com/trezor/trezor-firmware/commit/f9f91a2b71a82d0d369e2339c5f64c86c9660e18)? Without it the Cancel button on get_public_key confirm doesn't actually cancel the operation.

sure, LGTM.
(i mean it is not meaningfully changing anything if the Cancel does not work here but, well)

**Comment by mmilata** [comment]
Rebased and fixed a typing issue exposed by bf562cfd4bc48f5e6ab4f3f7a4b4f3e63b6dfb59.

> (i mean it is not meaningfully changing anything if the Cancel does not work here but, well)

The Cancel *works* in current master but was broken in this branch which was fixed. Or am I misunderstanding you?

**Comment by matejcik** [comment]
it works, but it's meaningless on a GetPublicKey message, because a potential attacker can obtain the pubkey silently anyway. I suppose we should remove the cancel button in the redesign.

**Pull request merged**
